### ISSUE TITLE
The format of .simg

### ISSUE THREAD

**Issue by elasolova** [alfonsoros88/RGBD-RF#1]
It will be really nice if you can provide an example input file. I tried to understand it from the code, but I couldn't. In the code it is stated that it is similar to Yale format, but I think it is not exactly Yale format.

Thanks,
elasolova

**Comment by rosds** [comment]
Hi elasolova,

I am sorry, the code is quite old and I am recently started to "modernize" and clean it a little bit. If you address me the part of the code you need I may be able to help you.

Regards,
Alfonso

**Comment by elasolova** [comment]
This part in Image.cpp
    // Read the image aspect ratio
    fscanf(fp, "%hd %hd;\n", &width, &height);

```
// Read image pixel labels
while ( fscanf(fp, "%d", &s) > 0 ) {
    l = s;
    pixelLabels.push_back(l);
}
fseek(fp, 1, SEEK_CUR);

// Read image depth labels
while ( fscanf(fp, "%d", &d) > 0 ) {
    pixelDepths.push_back(d);
}
fseek(fp, 1, SEEK_CUR);

// Read the vector I info
while ( fscanf(fp, "%d", &s) > 0 ) {
    I.push_back(s);
}
fseek(fp, 1, SEEK_CUR);

// Read the vector J info
while ( fscanf(fp, "%d", &s) > 0 ) {
    J.push_back(s);
}
```

It will be nice if you can specify what image labels, depth labels, I and J stand for. If there is an example .simg file, I can figure it  out myself, but from this snippet it is hard to understand the format for I,J, etc. 

You have the comment in Image.h if it would me helpful to remind you.
/**
-  @class TrainImage
  *
-  @brief This class is designed to contain an training image (i.e, This
-  kind of image contain the information of the depth and the labels).
-  The information is stored in something like the Yale Format for the
-  sparse matrixes.
  *
-  This images are used exclusively to train the random forest
-  algorithm. In our case we have preprocessed the real images formats
-  lile (bmp and exr) and write them in simple plane text files so we
-  can read them directly py the constructor.
  */

I have png files as input now. If I understand the format used, I will be able to convert them to specified .simg format.

Thanks for your fast reply,
elasolova

**Comment by rosds** [comment]
Right! I was just checking that code today. Sorry for the mess haha soon is going to be better.

So I am afraid that I don't have those *.simg files with me anymore but the idea was kind of simple. These *.simg files contained the training data for the random forest in a sparse matrix representation. We did it this way because we thought that the training data would take a lot of space.

Recalling the Yale representation http://en.wikipedia.org/wiki/Sparse_matrix#Yale, you have 3 vector to represent a sparse matrix (A, IA and JA). The values of the matrix are contained in the A vector. Back to the simg file, we have two "A" vectors, one containing the labels and an other with the depths of each pixel.  The vectors I and J are common for these two.

What you need to do then is to store the vector of labels and then the depth values for the non-zero pixels of your png files.

I hope this helps, let me know if you need any help

Best,
Alfonso

**Comment by elasolova** [comment]
Ok thanks. I was able to load a toy file now. I guess I need to figure out the rest. :) If I go successfully, maybe I can contribute to the hub here (maybe writing a read me, etc.) 

Best,
elasolova

**Comment by elasolova** [comment]
Sorry to ask many questions. But, I am trying to run the code on real data. Can you shortly describe the workflow for inputting? From what I understand, "pixel labels" correspond to body part labels. So, every pixel that has depth should be labeled as a body part. But, then what is meant by "depth labels"? So depth values should be divided in ranges and be labeled? I am trying really hard to grasp the input. If this is correct, then for test images we should not have any pixel labels and the forest has to label the body parts for us. If you provide a general overview of the workflow, I will be really grateful.

Best

**Comment by rosds** [comment]
Hi there,

yeah, I think that "depth labels" is an error. There are only depth values and labels. Each training pixel should have a depth value and a label associated to it, both are integers. 

Regards

**Comment by elasolova** [comment]
Hello again,

I have one more question. Does the split int nodes consider the x y position of pixels or only the depth value? In my experimentation I see that only depth is considered. Is that right?

Thanks,
elasolova

**Comment by rosds** [comment]
Hi!

Yes, it only considers the depth. I know, it's kind of odd when the name of
the project is RGBD.

The split candidates work over the pixels, they consist of an offset and a
threshold. They obtain the depth values for the offset positions from the
pixel, calculate the magnitude of those depth with respect to the pixel and
then simply threshold the difference to determine whether the traverse of
the tree should go right or left.

Regards,
Alfonso

**Comment by elasolova** [comment]
Yeah I see that is done in calcFeature. How will it be possible to integrate x or y(pixel position) into the feature.(because I think that will increase the accuracy a lot) Will be it be enough to change calcFeature function or I need to do additional modifications?

In an example I made calcFeature function to return x position of pixel. But,I did not get sensible results. Idk, maybe I am not still familiar with the code.

Thanks,
elasolova

**Comment by rosds** [comment]
Hi,

I am not sure if I understand what you mean by the (x, y) position of the
pixel. The whole idea for these split candidate is that they are really
computationally cheap but they are not good enough for a final
classification. So you hope that plug-in a bunch of them together is going
to produce a more powerful classification. This is also the same principle
as boosting.

In this case, each of these split candidates simply threshold the
difference of the depth between the pixel and two points in his
surrounding. That was the original idea of this paper
http://research.microsoft.com/pubs/145347/BodyPartRecognition.pdf. Of
course you can in principle change this split candidate and consider a
different feature. The only constraint is that these guys need to give LEFT
or RIGHT ans an answer to be able to connect them in a shape of a binary
tree. So you are free to try different splitting criteria and hopefully you
can find a better one :D

Best,

**Comment by elasolova** [comment]
Hello,

Thanks for clear explanation above.

One more issue. So I am trying to use offsets now. However, sometimes the getIndex method of Image class gets stuck. I suspect function lower_bound runs infinitely. Does this point to a problem in input format? It should return default depth perhaps even if it cannot find a pixel associated. What can be the problem?

Best,
elasolova

**Comment by elasolova** [comment]
So I finally did it. The whole thing works very well now. There was a format problem in input.
Plus, I got an accuracy of %74, with little training. Really nice code base. I appreciate your code. 
I will be happy to contribute in any way.

Thanks a lot,
elasolova

**Comment by rosds** [comment]
I am happy to hear (actually read) that! good job! The code is not that
great but I would love to modernize it to the c++11 standard, plus simplify
it a bit and some bug fixing.

But for that I need to find some time. If you want, we could work on that
together.

Best,
Alfonso

**Comment by elasolova** [comment]
Yeah, sure. I will be happy to work on this now that I am familiar with the code. I am not that expert when it comes to c++, but I will help as much as I can. For example, I wrote a matlab script to convert given png depth images and joint positions to required format. I can provide such utility tools at least. We can see what I can do.

Best,
elasolova

**Comment by etrigger** [comment]
Hi alfonsoros88,

Cool work and nice implementation.

I  just want to see if you have some training sample like depth images and corresponding labels. Because I don't quite understand how should I label the pixel. 

I see the paper you shared (http://research.microsoft.com/pubs/145347/BodyPartRecognition.pdf). They partition the body into 31 parts. Is that mean we would have 31 labels in total assigned  to each pixel of the foreground object in a depth image? In other words, the number of training label for one frame is equal to the number of pixels of a foreground object in a depth image.
![1](https://cloud.githubusercontent.com/assets/3329503/8391002/20a11e08-1ce2-11e5-9824-c6e91a19b960.png)

If that's true, is there any to speed up the manually annotation process as this would take me a long time to create all the training data.

My email is triggerliu@gmail.com.

Best regards,
Zhiguang

**Comment by rosds** [comment]
Hi Zhiguang,

right, I know there is some lack of documentation in this project, sorry
for that. Ehm.. if I remember correctly, we had png images with the label
information. Each pixel in these images had a value between 0-5 meaning, 0
is no label or background, 1 head, 2 left arm and so on. That means that we
only trained to classify 5 body parts (head, torso, left/right arm,
left/right leg). The reason for this is that the number of labels obviously
affect greatly on the training time and we didn't have such enormous
cluster as the people from Microsoft.

Anyway, we created these label png images in two ways, an manual way and an
artificial way. In the manual way, we simply took a bunch of depth images
from the Kinect and then color them by hand in gimp. This of course takes a
lot of time. For the artificial set, we got a 3d model of some resident
evil character and generated the images using Maya. This is way easier to
produce a lot of data in less time.

I hope this helps :D I am slowly re-factoring the code on my free time. I
will try to clarify these details in the documentation and make it more
flexible in the input data.

Best regards,
Alfonso

**Comment by etrigger** [comment]
Hi Alfonso, 
Thanks for you kind reply. I now understand the training data. I still have some questions:
(1) Do you think it's OK if I want to partition the foreground object into 20 parts using your project?
(2) I think the training process should start from this function "void rdf::RandomForest::trainForest(trainParams& tparams)", right? I don't familiar with these tricks in Linux as I always use Visual Studio, but I think there should be a main function that calling functions like training and predicting. Or I should write them myself? I found you have some command usage in ParseTreeArgs.cpp, but I don't know how to put them together. Can you give some hint?

Best regards,
Zhiguang.

**Comment by rosds** [comment]
Hi Yhiguang,

(1) In principle, yes. You can set the number of labels you want to learn,
but I wouldn't know how much would this affect you in the computation. For
us with 5 label, it took about a day using around 30 computers.

(2) Right, so the code is basically a library. The main class is
RandomForest, you need to instantiate this class in your program and link
your code to the library in order to use it. There are no binaries in the
code, it is ment to be used within some program. But it should be easy
though.

Let me know if you need some more help. Maybe it would be great to add some
demo or something for new users.

Best,
Alfonso

**Comment by etrigger** [comment]
Hi Alfonso,

Thanks for your reply. A simple demo would be great! I have two more questions:
(1) I think I only have 300 training samples or even fewer, would it take that long time? BTW, I guess I can only use one PC.(As I still can't run the code now, I just want to estimate the overall  run time)
(2) I do not how to set some of the parameters as follow in the  parameter file, it would be good if you can offer some init values from them.
offsetStart:
offsetEnd:
rangeStart:
rangetEnd:
treeNum:
labelNum: I guess this would be the number of labels we use, e.g. 5 in this project
maxDepth:
minSampleCount:
samplePixelNum:
offsetNum:
thresholdNum:
